parsel/selector.py

```diff
@@ -90,18 +112,23 @@ class SelectorList(List[_SelectorType]):
     """
 
     @typing.overload
-    def __getitem__(self, pos: int) -> _SelectorType:
+    def __getitem__(self, pos: "SupportsIndex") -> _SelectorType:
         pass
 
     @typing.overload
     def __getitem__(self, pos: slice) -> "SelectorList[_SelectorType]":
         pass
 
     def __getitem__(
-        self, pos: Union[int, slice]
+        self, pos: Union["SupportsIndex", slice]
     ) -> Union[_SelectorType, "SelectorList[_SelectorType]"]:
         o = super().__getitem__(pos)
-        return self.__class__(o) if isinstance(pos, slice) else o
+        if isinstance(pos, slice):
```

Comment:
These casts show a bigger problem: SelectorList is written as a list that can only contain Selectors, which is not checked at the run time and so cannot be expressed in typing properly (in technical terms Lists are invariant, not covariant). As we have Selector.selectorlist_cls and even tests that you can inherit your own selector and selector lists classes I feel like this can be overhauled to use some modern metaprogramming but that's out of scope, not trivial and will probably change the API and behavior.